x-pack/libbeat/management/generate.go

```diff
@@ -0,0 +1,274 @@
+func injectStreamProcessors(expected *proto.UnitExpectedConfig, inputType string, streamIter int, stream map[string]interface{}) (map[string]interface{}, error) {
+	//1. start by "repairing" config to add any missing fields
+	// logic from datastreamTypeFromInputNode
+	procInputType := inputType
+	if expected.DataStream.Type != "" {
+		procInputType = expected.DataStream.Type
+	}
+
+	procInputNamespace := DefaultNamespaceName
+	if expected.DataStream.Namespace != "" {
+		procInputNamespace = expected.DataStream.Namespace
+	}
+
+	var processors = []interface{}{}
+
+	// the AST injects input_id at the input level and not the stream level,
+	// for reasons I can't understand, as it just ends up shuffling it around
+	// to individual metricsets anyway, at least on metricbeat
+	if expected.Id != "" {
+		inputId := generateAddFieldsProcessor(mapstr.M{"input_id": expected.Id}, "@metadata")
+		processors = append(processors, inputId)
+	}
+
+	procInputDataset := DefaultDatasetName
+	if expected.Streams[streamIter].DataStream.Dataset != "" {
+		procInputDataset = expected.Streams[streamIter].DataStream.Dataset
+	}
+
+	//2. Actually add the processors
+	// namespace
+	datastream := generateAddFieldsProcessor(mapstr.M{"dataset": procInputDataset,
+		"namespace": procInputNamespace, "type": procInputType}, "data_stream")
+	processors = append(processors, datastream)
+
+	// dataset
+	event := generateAddFieldsProcessor(mapstr.M{"dataset": procInputDataset}, "event")
+	processors = append(processors, event)
+
+	// source stream
+	streamID := expected.Streams[streamIter].Id
+	sourceStream := generateAddFieldsProcessor(mapstr.M{"stream_id": streamID}, "@metadata")
+	processors = append(processors, sourceStream)
+
+	// figure out if we have any existing processors
+	currentProcs, ok := stream["processors"]
+	if !ok {
+		stream["processors"] = processors
+	} else {
+		currentProcsList, ok := currentProcs.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("error creating list of existing processors, got: %#v", currentProcs)
+		}
+		stream["processors"] = append(processors, currentProcsList...)
+
+	}
+
+	return stream, nil
+}
+
+// ===========
+// Config Processors
+// ===========
+
+func generateAddFieldsProcessor(fields mapstr.M, target string) mapstr.M {
+	return mapstr.M{
+		"add_fields": mapstr.M{
+			"fields": fields,
+			"target": target,
+		},
+	}
+}
+
+// This generates an opaque config blob used by all the beats
+// This has to handle both universal config changes and changes specific to the beats
+// This is a replacement for the AST code that lived in V1
+func generateBeatConfig(unitRaw *proto.UnitExpectedConfig, agentInfo *client.AgentInfo) ([]*reload.ConfigWithMeta, error) {
+	// We aren't guaranteed a DataStream field from the config
+	if unitRaw.DataStream == nil {
+		unitRaw.DataStream = &proto.DataStream{
+			Namespace: DefaultNamespaceName,
+			Dataset:   DefaultDatasetName,
+		}
+	} else {
+		if unitRaw.DataStream.Namespace == "" {
```

Comment:
Now that I've seen the nil errors Blake found, you need to keep in mind that every field in proto3 is optional and may return the default value. `proto.UnitExpectedConfig{}` is a valid message where each field has the default value: nil pointers, empty strings, 0 integers, etc.

Having to do this type of check is so common that the protobuf compiler auto-generates helpers for accessing nested fields that are nil safe. For example [GetNamespace](https://github.com/elastic/elastic-agent-client/blob/b0328d2fe484f3ecfe6d52efdf18058431743e83/pkg/proto/elastic-agent-client.pb.go#L1300).

Typical usage would be to chain these together such that if any field is nil along the chain you get the default value of the last message back: `unitRaw.GetDataStream().GetNamespace()` will return `""` if both `DataStream` or `Namespace` or nil. It is best to reflexively use these accessors instead of directly indexing the fields as it is very easy to forget a manual nil check.

Reply:
Ah, I completely forgot about the `Get*` GRPC functions. My solution was just to add a check to initialize the fields if they're nil:
```
	if raw.Streams[iter].DataStream == nil {
			raw.Streams[iter].DataStream = &proto.DataStream{}
		}
```

hence why all the subsequent calls aren't checking for nil values. Not sure if one way is "better", but I'm kinda tempted to move over to the `Get*` functions.

Reply:
hmm, yah, think I prefer the GRPC methods.